examples/run_nlp_trainer.py

```diff
@@ -0,0 +1,350 @@
+    # From https://stackoverflow.com/questions/26430861/make-pythons-warnings-warn-not-mention-itself
+    return f'{category.__name__}: {message} (source: {filename}:{lineno})\n'
+
+''' Generate a multiprocessing queue to store queue_size GPU IDs. '''
+def init_cuda_queue(queue_size: int) -> mp.Queue:
+    cuda_envs = mp.Queue(queue_size)
+    cuda_envs_list = range(queue_size)
+    for e in cuda_envs_list:
+        cuda_envs.put(e)
+
+    return cuda_envs
+
+''' Set up a single GPU CUDA environment on initialization of a mp process pool. '''
+def init_cuda_env(cuda_envs) -> None:
+    try:
+        env=cuda_envs.get()
+        torch.cuda.set_device(env)
+    except (BrokenPipeError, IOError):
+        pass
+
+    # fake a single node world 
+    os.environ['WORLD_SIZE'] = "1"
+    os.environ['LOCAL_WORLD_SIZE'] = "1"
+
+''' List all the checkpoints in a given S3 bucket and folder. ''' 
+def get_all_s3_checkpoints(bucket_name: str, folder_name: str) -> List[str]:
+    # pulls AWS credentials from environment variables, assumes that aws_session_token is defined
+    session = boto3.Session()
+
+    s3 = session.resource('s3')
+    bucket = s3.Bucket(bucket_name)
+    ckpt_names = []
+    for obj in bucket.objects.all():
+        if folder_name in obj.key:
+            obj_name = obj.key
+            ckpt_names.append(obj_name)
+
+    return ckpt_names
+
+''' Consolidate and prettify metrics. '''
+def output_metrics() -> None:
+    tasks = GLUE_METRICS[list(GLUE_METRICS.keys())[0]].keys()
+    large_tasks = ['MNLI', 'QNLI', 'QQP', 'SST-2']
+    # init table headers 
+    headers = ['Checkpoint']
+    headers.extend([f"{task.upper()}" for task in tasks])
+    headers.extend(['GLUE-Large', 'GLUE-All'])
+    tb = [headers]
+
+    # fill table 
+    for ckpt in GLUE_METRICS.keys():
+        output_line = [ckpt]
+        glue_all = 0
+        glue_large = 0
+        # Per task score
+        for task in sorted(GLUE_METRICS[ckpt].keys()):
+            output_line.append("{:.2f}".format(GLUE_METRICS[ckpt][task]))
+            glue_all += GLUE_METRICS[ckpt][task]
+            if task in large_tasks:
+                glue_large += GLUE_METRICS[ckpt][task]
+        # GLUE Large and GLUE All
+        output_line.append("{:.2f}".format(glue_large/len(large_tasks)))
+        output_line.append("{:.2f}".format(glue_all/len(tasks)))
+        tb.append(output_line)
+    
+    print(tabulate(tb, headers='firstrow'))
+
+''' Set up CUDA environment and process pool for given finetuning jobs. '''
+def setup_finetuning_jobs(finetune_tasks: List[str], save_ckpts: List[bool], ckpt_load_path: str, ckpt_filename: str,
+                                         save_folder: str, save_locally: bool, bucket: Optional[str] = None, parent_ckpt: Optional[str] = None, load_ignore_keys: Optional[List[str]] = None) -> None:
+    cuda_envs = init_cuda_queue(len(finetune_tasks))
```

Comment:
What if we have more finetune tasks than gpus available?

Reply:
The entrypoint was built based off the assumption that we're running an 8GPU job for both pretraining and finetuning. How high of a priority is it to support other use cases? Doing so would likely require a more formal job/resource queue.

Reply:
It seems like a good goal to make a tool like this _not_ be overly resource-specific. But in the meantime it seems like a good idea to have the code verify that the available resources satisfy the requirements and raise the appropriate error(s) if they don't.

Reply:
Discussed offline. Changed by using the number of current devices to initialize the Pool.